Create TTC Functions for Multi-LLM Generation

dnandakumar-nv opened the pull request:
## Description
This PR adds new test time compute (TTC) search and selection strategies, as well as a new function that allows users to generate candidate responses to a query using multiple LLMs and then select the best candidate using a judge. Usage example: 

One could use a YAML config such as 

```yaml
llms:
  llama_3_3_70b:
    _type: nim
    model_name: meta/llama-3.3-70b-instruct
    temperature: 0.7
    max_tokens: 1024

  llama_3_1_405b:
    _type: nim
    model_name: meta/llama-3.1-405b-instruct
    temperature: 0.1
    max_tokens: 1024

ttc_strategies:
  multi_llm_search:
    _type: multi_llm_generation
    llms:
      - llama_3_3_70b
      - llama_3_1_405b
    generation_template: |
      You are a helpful AI assistant. Please answer the following query concisely and accurately.
      
      Query: {prompt}
      
      Answer:

  judge_select:
    _type: llm_judge_selection
    judge_llm: llama_3_1_405b
    selection_template: |
      You are an expert judge. Select the best response for the query "{original_prompt}".
      
      Candidates:
      {results}
      
      Respond with exactly "SELECTED ITEM: <number>".

workflow:
  _type: multi_llm_judge_function
  search_strategy: multi_llm_search
  selection_strategy: judge_select
```

And run the workflow: 

```bash
nat run --config_file example_yaml.yml --input "Write me a limerick about GPU compute"
```

And the workflow will generate a candidate response using each configured LLM and use the judge to pick the best response. 

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/NeMo-Agent-Toolkit/blob/develop/docs/source/resources/contributing.md).
- We require that all contributors "sign-off" on their commits. This certifies that the contribution is your original work, or you have rights to submit it under the same license, or a compatible license.
  - Any contribution which contains commits that are not Signed-Off will not be accepted.
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a Multi-LLM Judge Function to generate multiple candidate responses and select the best via a judge LLM.
  * Introduced a Multi-LLM generation strategy for concurrent generation across multiple LLMs.
  * Added an LLM Judge selection strategy to rank and choose outputs using a configured judge model.

* **Documentation**
  * Updated TTC reference to include the new generation, selection configs, and the orchestration function.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dnandakumar-nv committed:
Add LLM-based multi-strategy orchestration for search and selection

Introduced `multi_llm_generation` and `llm_judge_selection` strategies to support multi-LLM response generation and judge-based selection. Added `multi_llm_judge_function` for coordinated execution of these strategies. Enhanced model validation with `@classmethod` decorators across configurations.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor code for improved readability and styling

Reformatted long lines, combined multi-line expressions, and simplified code structure across multiple files for better readability and adherence to PEP 8. Maintained functionality while enhancing the overall code quality and consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge branch 'develop' into ttc-llm-council

dnandakumar-nv commented:
/ok to test 2b5177c83f53137db3760447249e690422696ed3

willkill07 commented on src/nat/experimental/test_time_compute/models/search_config.py:
```diff
@@ -114,7 +116,24 @@ class MultiQueryRetrievalSearchConfig(TTCStrategyBaseConfig, name="multi_query_r
+    @model_validator(mode="before")
+    @classmethod
+    def validate_config(cls, values: dict) -> dict:
+        if not values.get('llms'):
```
Need to also check that `llms` is a list

willkill07 reviewed (changes requested):
Could we add some user-facing documentation at minimum?

An example would also be nice, though IMO not required for merge.

dnandakumar-nv committed:
Merge branch 'develop' into ttc-llm-council

dnandakumar-nv committed:
Update LLM validation and documentation for multi-LLM features

Enhanced LLM validation to ensure `llms` is a non-empty list in `SearchConfig`. Updated documentation to include new strategies (`multi_llm_generation`, `llm_judge_selection`) and functions (`multi_llm_judge_function`) for consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test ca868cca92f6aed7cdf82e8423bcf7a8831c07fc

dnandakumar-nv commented:
/merge
